# Apartado B

In [1]:
import pandas as pd
import numpy as np
import datetime
import seaborn as sns

#Leemos la base de datos
data = pd.read_csv("./weatherAUS.csv")

## EDA

In [2]:
#Obtenemos la información general de la base de datos
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

In [3]:
data.head(10)

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   
5  2008-12-06   Albury     14.6     29.7       0.2          NaN       NaN   
6  2008-12-07   Albury     14.3     25.0       0.0          NaN       NaN   
7  2008-12-08   Albury      7.7     26.7       0.0          NaN       NaN   
8  2008-12-09   Albury      9.7     31.9       0.0          NaN       NaN   
9  2008-12-10   Albury     13.1     30.1       1.4          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   
5         WNW           56.0          W  ...        55.0         23.0   
6           W           50.0         SW  ...        49.0         19.0   
7           W           35.0        SSE  ...        48.0         19.0   
8         NNW           80.0         SE  ...        42.0          9.0   
9           W           28.0          S  ...        58.0         27.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   
5       1009.2       1005.4       NaN       NaN     20.6     28.9         No   
6       1009.6       1008.2       1.0       NaN     18.1     24.6         No   
7       1013.4       1010.1       NaN       NaN     16.3     25.5         No   
8       1008.9       1003.6       NaN       NaN     18.3     30.2         No   
9       1007.0       1005.7       NaN       NaN     20.1     28.2        Yes   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  
5            No  
6            No  
7            No  
8           Yes  
9            No  

[10 rows x 23 columns]

In [4]:
data.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  143975.000000  144199.000000  142199.000000  82670.000000   
mean       12.194034      23.221348       2.360918      5.468232   
std         6.398495       7.119049       8.478060      4.193704   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.900000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  75625.000000  135197.000000  143693.000000  142398.000000   
mean       7.611178      40.035230      14.043426      18.662657   
std        3.785483      13.607062       8.915375       8.809800   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.800000      31.000000       7.000000      13.000000   
50%        8.400000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm   Pressure9am    Pressure3pm  \
count  142806.000000  140953.000000  130395.00000  130432.000000   
mean       68.880831      51.539116    1017.64994    1015.255889   
std        19.029164      20.795902       7.10653       7.037414   
min         0.000000       0.000000     980.50000     977.100000   
25%        57.000000      37.000000    1012.90000    1010.400000   
50%        70.000000      52.000000    1017.60000    1015.200000   
75%        83.000000      66.000000    1022.40000    1020.000000   
max       100.000000     100.000000    1041.00000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am       Temp3pm  
count  89572.000000  86102.000000  143693.000000  141851.00000  
mean       4.447461      4.509930      16.990631      21.68339  
std        2.887159      2.720357       6.488753       6.93665  
min        0.000000      0.000000      -7.200000      -5.40000  
25%        1.000000      2.000000      12.300000      16.60000  
50%        5.000000      5.000000      16.700000      21.10000  
75%        7.000000      7.000000      21.600000      26.40000  
max        9.000000      9.000000      40.200000      46.70000

La base de datos tiene 145460 entradas con 23 atributos. El atributo "Date" es de tipo string, hay que convertirlo en datetime, porque significa una fecha. Los atributos "Location", "WindGustDir", "WindDir9am", "WindDir3pm", "RainToday" y "RainTomorrow" son las variables categóricas. Los demás atributos son numéricos. El "RainTomorrow" será el atributo objetivo, porque requiere una predicción de una de las dos categorías ("Yes" o "No").

In [5]:
categoricas = ["Location", "WindGustDir", "WindDir9am", "WindDir3pm", "RainToday", "RainTomorrow"]

In [6]:
data[categoricas] = data[categoricas].astype("category")
data[categoricas].describe()

Location WindGustDir WindDir9am WindDir3pm RainToday RainTomorrow
count     145460      135134     134894     141232    142199       142193
unique        49          16         16         16         2            2
top     Canberra           W          N         SE        No           No
freq        3436        9915      11758      10838    110319       110316

El atributo "Location" tiene entradas para 49 ciudades de Australia. "WindDirGust", "WindDir9am", "WindDir3am" pueden tener 16 valores diferentes.

In [7]:
data["Date"] = pd.to_datetime(data["Date"])
data["Date"].describe(datetime_is_numeric=True)

count                           145460
mean     2013-04-04 21:08:51.907053568
min                2007-11-01 00:00:00
25%                2011-01-11 00:00:00
50%                2013-06-02 00:00:00
75%                2015-06-14 00:00:00
max                2017-06-25 00:00:00
Name: Date, dtype: object

In [8]:
data["Day"] = data["Date"].dt.day
data["Month"] = data["Date"].dt.month
data["Year"] = data["Date"].dt.year
data.drop("Date", axis = 1, inplace = True)

El atributo objetivo no está blanceado (tiene 110316 entradas negativas, que es mucho más que las positivas)

## Preprocessing

Como hemos visto en el apartado anterior, los datos numéricos no están estandarizados. Hay que estandarizarlos para procesar los valores en el mismo rango (promedio = 0, desviación estándar = 1). Así el procesamiento será más rápido.

Algunos atributos tienen valores nulos. Los vamos a tratar de maneras diferentes, más adientes para cada tipo de dato en concreto.

Los valores nulos de atributos numéricos serán sustituidos por valores promedios de dichos atributos para cada par (Mes, Ciudad). Primero, construimos la tabla agrupada por ciudades y fefchas con valores promedios de atributos numéricos. Aprovechamos de que no tenemos valores nulos para localidades y fechas.

In [9]:
from sklearn.preprocessing import LabelEncoder
mean_features = ["Evaporation","MaxTemp","Rainfall","MinTemp","Sunshine","WindGustSpeed","WindSpeed9am","WindSpeed3pm",
                "Humidity9am","Humidity3pm","Pressure9am","Pressure3pm","Cloud9am","Cloud3pm","Temp9am","Temp3pm"]
mean_table = data.groupby(["Month", "Location"], as_index = False)[mean_features].mean()
mean_table.reset_index()
le = LabelEncoder()
mean_table["Location"] = le.fit_transform(mean_table["Location"])
data["Location"] = le.fit_transform(data["Location"])
mean_table

Month  Location  Evaporation    MaxTemp  Rainfall    MinTemp   Sunshine  \
0        1         0    11.885915  30.954122  0.769118  18.062007  11.178488   
1        1         1     7.088312  23.342336  0.884420  16.796283   8.521491   
2        1         2          NaN  32.925806  1.475269  16.874194        NaN   
3        1         3    12.558015  36.132014  2.194224  21.789606   9.996522   
4        1         4          NaN  30.598561  2.695588  17.543682        NaN   
..     ...       ...          ...        ...       ...        ...        ...   
583     12        44     7.117130  26.180645  2.289720  13.017674   8.753456   
584     12        45    10.149324  27.676037  2.935227  16.910599   7.093578   
585     12        46          NaN  24.976959  0.382243  12.422581        NaN   
586     12        47          NaN  23.761826  3.218182  17.472385        NaN   
587     12        48    16.496512  33.571889  0.385096  18.623041  11.060938   

     WindGustSpeed  WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  \
0        38.931900      9.935484     16.379061    47.273381    32.452899   
1              NaN     15.848837     24.109453    70.520446    69.403756   
2        39.064516     10.960573     15.881295    53.817204    29.501792   
3        45.748201     18.078853     19.374101    38.211470    26.892086   
4        36.805147      7.374545     14.882353    72.844765    49.776557   
..             ...           ...           ...          ...          ...   
583      39.855769     10.797235     16.287037    66.783410    45.603687   
584      44.052885     15.334884     24.847222    68.308756    54.949309   
585      42.465438     17.829493     23.124424    56.506912    50.202765   
586      46.668033     17.233607     22.991870    72.500000    70.929752   
587      49.685185     22.967742     20.620370    39.059908    20.231481   

     Pressure9am  Pressure3pm  Cloud9am  Cloud3pm    Temp9am    Temp3pm  
0    1013.725448  1011.937906       NaN       NaN  23.029496  29.314440  
1    1015.175940  1013.509398  5.191882  4.046729  20.467528  21.492991  
2    1012.535125  1009.727240  5.810345  4.740260  22.783513  30.996416  
3    1008.449104  1005.003957  3.585938  4.412451  29.235484  34.535612  
4    1014.144765  1011.391941       NaN       NaN  22.449819  28.737729  
..           ...          ...       ...       ...        ...        ...  
583  1013.865278  1012.015207  5.110599  4.838710  18.104608  24.214747  
584  1014.549770  1012.468664  5.390110  4.740331  22.368664  25.875576  
585  1015.718894  1014.177880       NaN       NaN  19.875576  22.564516  
586  1014.034855  1012.263968  6.224806  6.256198  20.700417  21.878099  
587  1012.987097  1010.519907  2.236686  2.454545  23.897235  31.401389  

[588 rows x 18 columns]

Como se ve, esta tabla también tiene valores nulos. Como esta segunda tabla tiene pocas filas, podemos utilizar el KNN Imputer.

In [10]:
from sklearn.impute import KNNImputer
knni = KNNImputer()
mean_table = pd.DataFrame(knni.fit_transform(mean_table),columns = mean_table.columns)
mean_table.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 588 entries, 0 to 587
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Month          588 non-null    float64
 1   Location       588 non-null    float64
 2   Evaporation    588 non-null    float64
 3   MaxTemp        588 non-null    float64
 4   Rainfall       588 non-null    float64
 5   MinTemp        588 non-null    float64
 6   Sunshine       588 non-null    float64
 7   WindGustSpeed  588 non-null    float64
 8   WindSpeed9am   588 non-null    float64
 9   WindSpeed3pm   588 non-null    float64
 10  Humidity9am    588 non-null    float64
 11  Humidity3pm    588 non-null    float64
 12  Pressure9am    588 non-null    float64
 13  Pressure3pm    588 non-null    float64
 14  Cloud9am       588 non-null    float64
 15  Cloud3pm       588 non-null    float64
 16  Temp9am        588 non-null    float64
 17  Temp3pm        588 non-null    float64
dtypes: float64

In [11]:
for feature in mean_features:
    for month in data["Month"].unique():
        for location in data["Location"].unique():
            mask = (data["Location"] == location) & (data["Month"] == month)
            mask_mean = (mean_table["Location"] == location) & (mean_table["Month"] == month)
            val = mean_table.loc[mask_mean, feature].item()
            data.loc[mask, feature] = data.loc[mask,feature].fillna(val)

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 25 columns):
 #   Column         Non-Null Count   Dtype   
---  ------         --------------   -----   
 0   Location       145460 non-null  int32   
 1   MinTemp        145460 non-null  float64 
 2   MaxTemp        145460 non-null  float64 
 3   Rainfall       145460 non-null  float64 
 4   Evaporation    145460 non-null  float64 
 5   Sunshine       145460 non-null  float64 
 6   WindGustDir    135134 non-null  category
 7   WindGustSpeed  145460 non-null  float64 
 8   WindDir9am     134894 non-null  category
 9   WindDir3pm     141232 non-null  category
 10  WindSpeed9am   145460 non-null  float64 
 11  WindSpeed3pm   145460 non-null  float64 
 12  Humidity9am    145460 non-null  float64 
 13  Humidity3pm    145460 non-null  float64 
 14  Pressure9am    145460 non-null  float64 
 15  Pressure3pm    145460 non-null  float64 
 16  Cloud9am       145460 non-null  float64 
 17  Cloud3pm  

Ahora es evidente que hemos llenado la mayor parte de la tabla. Ya se puede simplemente eliminar las filas que tengan valores nulos, porque no perdemos mucha información.

In [13]:
data = data.dropna()
data.reset_index()


index  Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0            0         2     13.4     22.9       0.6     7.239858  9.287474   
1            1         2      7.4     25.1       0.0     7.239858  9.287474   
2            2         2     12.9     25.7       0.0     7.239858  9.287474   
3            3         2      9.2     28.0       0.0     7.239858  9.287474   
4            4         2     17.5     32.3       1.0     7.239858  9.287474   
...        ...       ...      ...      ...       ...          ...       ...   
123705  145454        41      3.5     21.8       0.0     4.391246  7.953688   
123706  145455        41      2.8     23.4       0.0     4.391246  7.953688   
123707  145456        41      3.6     25.3       0.0     4.391246  7.953688   
123708  145457        41      5.4     26.9       0.0     4.391246  7.953688   
123709  145458        41      7.8     27.0       0.0     4.391246  7.953688   

       WindGustDir  WindGustSpeed WindDir9am  ... Pressure3pm  Cloud9am  \
0                W           44.0          W  ...      1007.1  8.000000   
1              WNW           44.0        NNW  ...      1007.8  5.539683   
2              WSW           46.0          W  ...      1008.7  5.539683   
3               NE           24.0         SE  ...      1012.8  5.539683   
4                W           41.0        ENE  ...      1006.0  7.000000   
...            ...            ...        ...  ...         ...       ...   
123705           E           31.0        ESE  ...      1021.2  5.714286   
123706           E           31.0         SE  ...      1020.3  5.714286   
123707         NNW           22.0         SE  ...      1019.1  5.714286   
123708           N           37.0         SE  ...      1016.8  5.714286   
123709          SE           28.0        SSE  ...      1016.5  3.000000   

        Cloud3pm  Temp9am  Temp3pm  RainToday  RainTomorrow  Day  Month  Year  
0       4.901099     16.9     21.8         No            No    1     12  2008  
1       4.901099     17.2     24.3         No            No    2     12  2008  
2       2.000000     21.0     23.2         No            No    3     12  2008  
3       4.901099     18.1     26.5         No            No    4     12  2008  
4       8.000000     17.8     29.7         No            No    5     12  2008  
...          ...      ...      ...        ...           ...  ...    ...   ...  
123705  4.500000      9.4     20.9         No            No   20      6  2017  
123706  4.500000     10.1     22.4         No            No   21      6  2017  
123707  4.500000     10.9     24.5         No            No   22      6  2017  
123708  4.500000     12.5     26.1         No            No   23      6  2017  
123709  2.000000     15.1     26.0         No            No   24      6  2017  

[123710 rows x 26 columns]

In [14]:
le = LabelEncoder()
for cat in categoricas:
    data[cat] = le.fit_transform(data[cat])

In [15]:
data

Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0              1     13.4     22.9       0.6     7.239858  9.287474   
1              1      7.4     25.1       0.0     7.239858  9.287474   
2              1     12.9     25.7       0.0     7.239858  9.287474   
3              1      9.2     28.0       0.0     7.239858  9.287474   
4              1     17.5     32.3       1.0     7.239858  9.287474   
...          ...      ...      ...       ...          ...       ...   
145454        39      3.5     21.8       0.0     4.391246  7.953688   
145455        39      2.8     23.4       0.0     4.391246  7.953688   
145456        39      3.6     25.3       0.0     4.391246  7.953688   
145457        39      5.4     26.9       0.0     4.391246  7.953688   
145458        39      7.8     27.0       0.0     4.391246  7.953688   

        WindGustDir  WindGustSpeed  WindDir9am  WindDir3pm  ...  Pressure3pm  \
0                13           44.0          13          14  ...       1007.1   
1                14           44.0           6          15  ...       1007.8   
2                15           46.0          13          15  ...       1008.7   
3                 4           24.0           9           0  ...       1012.8   
4                13           41.0           1           7  ...       1006.0   
...             ...            ...         ...         ...  ...          ...   
145454            0           31.0           2           0  ...       1021.2   
145455            0           31.0           9           1  ...       1020.3   
145456            6           22.0           9           3  ...       1019.1   
145457            3           37.0           9          14  ...       1016.8   
145458            9           28.0          10           3  ...       1016.5   

        Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RainTomorrow  Day  \
0       8.000000  4.901099     16.9     21.8          0             0    1   
1       5.539683  4.901099     17.2     24.3          0             0    2   
2       5.539683  2.000000     21.0     23.2          0             0    3   
3       5.539683  4.901099     18.1     26.5          0             0    4   
4       7.000000  8.000000     17.8     29.7          0             0    5   
...          ...       ...      ...      ...        ...           ...  ...   
145454  5.714286  4.500000      9.4     20.9          0             0   20   
145455  5.714286  4.500000     10.1     22.4          0             0   21   
145456  5.714286  4.500000     10.9     24.5          0             0   22   
145457  5.714286  4.500000     12.5     26.1          0             0   23   
145458  3.000000  2.000000     15.1     26.0          0             0   24   

        Month  Year  
0          12  2008  
1          12  2008  
2          12  2008  
3          12  2008  
4          12  2008  
...       ...   ...  
145454      6  2017  
145455      6  2017  
145456      6  2017  
145457      6  2017  
145458      6  2017  

[123710 rows x 25 columns]

In [16]:
for col in data.columns:
    data[col] = (data[col] - data[col].mean()) / data[col].std()

In [17]:
data.describe()

Location       MinTemp       MaxTemp      Rainfall   Evaporation  \
count  1.237100e+05  1.237100e+05  1.237100e+05  1.237100e+05  1.237100e+05   
mean   1.613228e-14  2.033685e-14  1.444951e-14 -5.084136e-14  4.972928e-14   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -1.671183e+00 -3.283946e+00 -3.922950e+00 -2.785706e-01 -1.432131e+00   
25%   -8.633248e-01 -7.247995e-01 -7.459979e-01 -2.785706e-01 -7.311149e-01   
50%    1.797543e-02 -3.398705e-02 -6.621337e-02 -2.785706e-01 -1.395829e-01   
75%    8.258340e-01  7.353268e-01  7.106832e-01 -1.847624e-01  4.528349e-01   
max    1.707134e+00  3.372974e+00  3.415948e+00  4.282630e+01  2.177988e+01   

           Sunshine   WindGustDir  WindGustSpeed    WindDir9am    WindDir3pm  \
count  1.237100e+05  1.237100e+05   1.237100e+05  1.237100e+05  1.237100e+05   
mean  -1.080053e-13 -3.715418e-15   2.197064e-14 -2.730997e-14 -8.835243e-15   
std    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  1.000000e+00   
min   -2.505731e+00 -1.656448e+00  -2.515961e+00 -1.597242e+00 -1.683232e+00   
25%   -5.721887e-01 -1.013354e+00  -7.233956e-01 -9.380654e-01 -8.171034e-01   
50%    3.724726e-02  5.846988e-02  -1.258739e-01 -5.916304e-02  4.902470e-02   
75%    7.279516e-01  9.159288e-01   5.463380e-01  8.197394e-01  9.151528e-01   
max    2.328124e+00  1.559023e+00   7.044386e+00  1.698642e+00  1.564749e+00   

       ...   Pressure3pm      Cloud9am      Cloud3pm       Temp9am  \
count  ...  1.237100e+05  1.237100e+05  1.237100e+05  1.237100e+05   
mean   ...  8.584277e-13 -5.635778e-14 -1.118720e-13 -3.231287e-14   
std    ...  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min    ... -5.631085e+00 -1.956519e+00 -2.113799e+00 -3.761945e+00   
25%    ... -6.664063e-01 -6.873390e-01 -7.476461e-01 -7.308497e-01   
50%    ...  1.257576e-02  1.587814e-01  1.631229e-01 -5.385373e-02   
75%    ...  6.377481e-01  7.748163e-01  6.332765e-01  7.154599e-01   
max    ...  3.631375e+00  1.851022e+00  1.984661e+00  3.531148e+00   

            Temp3pm     RainToday  RainTomorrow           Day         Month  \
count  1.237100e+05  1.237100e+05  1.237100e+05  1.237100e+05  1.237100e+05   
mean   3.543043e-14 -1.242964e-14 -2.015102e-14 -4.279586e-16 -1.415833e-15   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -3.867274e+00 -5.378301e-01 -5.332813e-01 -1.673676e+00 -1.563064e+00   
25%   -7.426192e-01 -5.378301e-01 -5.332813e-01 -8.772172e-01 -9.875433e-01   
50%   -6.396109e-02 -5.378301e-01 -5.332813e-01  3.302106e-02 -1.242620e-01   
75%    6.995293e-01 -5.378301e-01 -5.332813e-01  8.294795e-01  7.390193e-01   
max    3.498994e+00  1.859308e+00  1.875168e+00  1.739718e+00  1.602301e+00   

               Year  
count  1.237100e+05  
mean   4.661218e-14  
std    1.000000e+00  
min   -2.296291e+00  
25%   -7.143090e-01  
50%    7.668222e-02  
75%    8.676735e-01  
max    1.658665e+00  

[8 rows x 25 columns]

Teóricamente podríamos aplicar el PCA, pero no lo vamos a hacer. El PCA "eliminaría" variables que pudiesen ser linealmente dependientes y comprimiría la base de datos. Pero como en nustra base no hay variables "evidentemente" dependientes, aprovecharemos de ello para sacar más información del dataset.

Tampoco vamos a utilizar PolynomialFeatures, porque este método sirve para aumentar el número de las variables (las tenemos bastantes) o para construir los atributos que tengan sentido (no tiene sentido multiplicar, por ejemplo, "Sunshine" por "Evaporation")

## Model selection

Considerando el tamaño de nustra base de datos se puede eligir los siguientes modelos: Kernel SVM, Multilayer Perceptron, Random Forest, Gradient Boosting. Suponemos que serán más precisos. Pero tambien se puede considerar tambien los árboles de decisión, regresión logística y Naïve Bayes, porque se ajustarán más rápido.

Hemos decidido usar el RBF kernel, porque suponemos que la distribución de clases de nuestra variable objetivo no puede ser lineal (la probabilidad de la lluvia está compuesta por muchas circunstancias que no dependen una de otra).

Los modelos ensemble seguramente serán precisos, pero tardarán bastante, ya que se usan mas para datasets más pequeños

## Crossvalidation

La cross-validación de los resultados es la parte más importante en la construcción de un buen modelo, que puede dar buenos resultados en el conjnunto de entrenamiento, pero fracasará en casos no vistos. Cross-validación no ayudará a evitar el overfitting y el underfitting.

La mejor estrategia para separar el dataset en conjuntos de entrenamiento y validación es guardar más datos en el train set, porque así utilizamos casi todos los datos disponibles. Pero no es posible evitar el test-set, porque sin él la probabilidad de overfitting es mucho mayor. Nuestra base de datos es bastante grande para poder probar separar el dataset en 90/10 de train y test, respectivamente.

También, se puede usar la estrategia K-Fold con mayor K posible y con mayor número de repeticiones. De esta manera el tipo más interesante puede ser el Repeated Stratified K-Fold. Pero tardará bastante. Habrá que probar diferentes tipos.

En nuestro caso seguramente no podemos utilizar la estrategia Leave-One-Out, ya que así tendrémos más de 100 mil train sets, que es demasiado costoso.

## Metric Analysis

Para evaluar la calidad de nuestro clasificador no podemos utilizar accuracy, porque los datos no están balanceados. Con tantos valores negativos lo que necesitamos es F1-Score, porque es una combinación de Precision y Recall.
Tambien necesitamos AUC.

La Precisión-Recall Curve y la ROC Curve serán igualmente relevantes para nustra base de datos, porque nos dan toda la información desde puntos de vista diferentes. 
PRC es la curva que muestra la relación entre Precision y Recall. ROC muestra la relación entre False Positive Rate y True Positive Rate.
